docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+
+* ``'ARROW:extension:name'`` for the string name identifying the
+  custom data type. We recommend that you use a "namespace"-style
+  prefix for extension type names to minimize the possibility of
+  conflicts with multiple Arrow readers and writers in the same
+  application. For example, use ``myorg.name_of_type`` instead of
+  simply ``name_of_type``
+* ``'ARROW:extension:metadata'`` for a serialized representation
+  of the ``ExtensionType`` necessary to reconstruct the custom type
+
+This extension metadata can annotate any of the built-in Arrow logical
+types. The intent is that an implementation that does not support an
+extension type can still handle the underlying data. For example a
+16-byte UUID value could be embedded in ``FixedSizeBinary(16)``, and
+implementations that do not have this extension type can still work
+with the underlying binary values and pass along the
+``custom_metadata`` in subsequent Arrow protocol messages.
+
+Extension types may or may not use the
+``'ARROW:extension:metadata'`` field. Let's consider some example
+extension types:
+
+* ``uuid`` represented as ``FixedSizeBinary(16)`` with empty metadata
+* ``latitude-longitude`` represented as ``struct<latitude: double,
+  longitude: double>``, and empty metadata
+* ``tensor`` (multidimensional array) stored as ``Binary`` values and
+  having serialized metadata indicating the data type and shape of
+  each value. This could be JSON like ``{'type': 'int8', 'shape': [4,
+  5]}`` for a 4x5 cell tensor.
+* ``trading-time`` represented as ``Timestamp`` with serialized
+  metadata indicating the market trading calendar the data corresponds
+  to
+
+Implementation guidelines
+=========================
+
+An execution engine (or framework, or UDF executor, or storage engine,
+etc) can implement only a subset of the Arrow spec and/or extend it
+given the following constraints:
+
+Implementing a subset the spec
+------------------------------
+
+* **If only producing (and not consuming) arrow vectors**: Any subset
+  of the vector spec and the corresponding metadata can be implemented.
+* **If consuming and producing vectors**: There is a minimal subset of
+  vectors to be supported.  Production of a subset of vectors and
+  their corresponding metadata is always fine.  Consumption of vectors
+  should at least convert the unsupported input vectors to the
+  supported subset (for example Timestamp.millis to timestamp.micros
+  or int32 to int64).
+
+Extensibility
+-------------
+
+An execution engine implementor can also extend their memory
+representation with their own vectors internally as long as they are
+never exposed. Before sending data to another system expecting Arrow
+data, these custom vectors should be converted to a type that exist in
+the Arrow spec.  An example of this is operating on compressed data.
+These custom vectors are not exchanged externally and there is no
+support for custom metadata.
+
+Integration Testing
```

Comment:
I would put this in a separate document as it's really not important for understanding the data format.